pkg/destroy/vsphere/vsphere.go

```diff
@@ -52,98 +63,127 @@ func New(logger logrus.FieldLogger, metadata *installertypes.ClusterMetadata) (p
-				if err != nil {
-					return err
-				}
-				task.Wait(ctx)
-				virtualMachineLogger.Debug("Powered off")
-			}
+func isNotFound(err error) bool {
+	return err != nil && strings.HasSuffix(err.Error(), http.StatusText(http.StatusNotFound))
+}
 
-			task, err := vm.Destroy(ctx)
-			if err != nil {
-				return err
-			}
-			task.Wait(ctx)
-			virtualMachineLogger.Info("Destroyed")
-		}
-	}
-	return nil
+func (o *ClusterUninstaller) contextWithTimeout() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(o.context, defaultTimeout)
 }
-func deleteFolder(ctx context.Context, client *vim25.Client, folderMoList []mo.Folder, logger logrus.FieldLogger) error {
-	ctx, cancel := context.WithTimeout(ctx, time.Second*60)
-	defer cancel()
 
-	// If there are no children in the folder go ahead an remove it
-	if len(folderMoList[0].ChildEntity) == 0 {
-		folderLogger := logger.WithField("Folder", folderMoList[0].Name)
+func (o *ClusterUninstaller) getAttachedObjectsOnTag() ([]tags.AttachedObjects, error) {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
 
-		folder := object.NewFolder(client, folderMoList[0].Reference())
-		task, err := folder.Destroy(ctx)
-		if err != nil {
-			return err
-		}
-		task.Wait(ctx)
-		folderLogger.Info("Destroyed")
-	} else {
-		return errors.Errorf("Expected Folder %s to be empty", folderMoList[0].Name)
+	o.Logger.Debug("Find attached objects on tag")
+	tagManager := tags.NewManager(o.RestClient)
+	attached, err := tagManager.GetAttachedObjectsOnTags(ctx, []string{o.InfraID})
+	if err != nil {
+		return nil, err
 	}
 
-	return nil
+	return attached, nil
 }
-func getFolderManagedObjects(ctx context.Context, client *vim25.Client, moRef []types.ManagedObjectReference) ([]mo.Folder, error) {
+
+func (o *ClusterUninstaller) getFolderManagedObjects(moRef []types.ManagedObjectReference) ([]mo.Folder, error) {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
 	var folderMoList []mo.Folder
-	pc := property.DefaultCollector(client)
+	pc := property.DefaultCollector(o.Client)
 	err := pc.Retrieve(ctx, moRef, nil, &folderMoList)
 	if err != nil {
 		return nil, err
 	}
 	return folderMoList, nil
 }
-func getVirtualMachineManagedObjects(ctx context.Context, client *vim25.Client, moRef []types.ManagedObjectReference) ([]mo.VirtualMachine, error) {
-	var virtualMachineMoList []mo.VirtualMachine
 
-	pc := property.DefaultCollector(client)
-	err := pc.Retrieve(ctx, moRef, nil, &virtualMachineMoList)
-	if err != nil {
+func (o *ClusterUninstaller) listFolders() ([]mo.Folder, error) {
+	tagAttachedObjects, err := o.getAttachedObjectsOnTag()
+	if err != nil && !isNotFound(err) {
 		return nil, err
 	}
-	return virtualMachineMoList, nil
+
+	var folderList []types.ManagedObjectReference
+	for _, attachedObject := range tagAttachedObjects {
+		for _, ref := range attachedObject.ObjectIDs {
+			if ref.Reference().Type == "Folder" {
+				folderList = append(folderList, ref.Reference())
+			}
+		}
+	}
```

Comment:
Can you add the filtering on type to the `getAttachedObjectsOnTag` function so that we are not duplicating this logic?